### PR TITLE
1st commit. Auth and Reg

### PR DESCRIPTION
**Pull request by Main444**
Added Registration and Authentication

### CODE REVIEW

File: src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.csrf().disable();
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java:23]: enable csrf

File: src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        http.formLogin().loginPage("/login-page").defaultSuccessUrl("/main-page",true);
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java]: duplicate code, make it inline

File: src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        http.authorizeRequests().mvcMatchers("/login-page","/registration").anonymous();
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java]: create security proprties file, and use it instead hardcode

File: src/main/java/com/pharmacy/optican/demo/controller/PublicController.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    private final UserService userService;
+
+    @Autowired
+    public PublicController(UserService userService) {
+        this.userService = userService;
+    }
+
+
+    @GetMapping("/registration")
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/PublicController.java]: make couple diferent controller

File: src/main/java/com/pharmacy/optican/demo/controller/PublicController.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        return "registration-page";
+    }
+
+    @PostMapping("/registration")
+    public String addUser(User user){
+        userService.saveUser(user);
+        return "redirect:/main-page";
+    }
+
+    @GetMapping("/main-page")
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/PublicController.java]: use _ instead -

File: src/main/java/com/pharmacy/optican/demo/controller/PublicController.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    public String mainPage(){
+        return "main-page";
+    }
+
+    @GetMapping("/login-page")
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/PublicController.java]: the same

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    private final UserService userService;
+    private final PasswordEncoder passwordEncoder;
+
+    @Autowired
+    public UserController(UserService userService, PasswordEncoder passwordEncoder) {
+        this.userService = userService;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    @GetMapping("/user-page")
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: _ instead -

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    public String userPage(Authentication authentication, Model model){
+        model.addAttribute("user",userService.findUserByName(authentication.getName()).get());
+        return "user-page";
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/UserController.java:26]: _ instead -

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @PostMapping("/update-user")
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: _ instead -

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+
+import java.util.Optional;
+
+@Controller
+public class UserController {
+
+
+    private final UserService userService;
+    private final PasswordEncoder passwordEncoder;
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: remove passwordEncoder from controller

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    @GetMapping("/user-page")
+    public String userPage(Authentication authentication, Model model){
+        model.addAttribute("user",userService.findUserByName(authentication.getName()).get());
+        return "user-page";
+    }
+
+    @PostMapping("/update-user")
+    public String userUpdate(User user){
+        Optional<User> oldUser = userService.findUserByName(user.getName());
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: return user from service instead Optional<user>

File: src/main/java/com/pharmacy/optican/demo/model/User.java [hunk 12]
```diff
@@ -16,9 +16,14 @@
     @Column(name = "id")
     @GeneratedValue(strategy = GenerationType.SEQUENCE,generator = "user_seq")
     @SequenceGenerator(name = "user_seq", sequenceName = "user_id_seq", allocationSize = 1)
-    private Long id;
+    private Integer id;
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/model/User.java]: Use Long instead Integer

File: src/main/java/com/pharmacy/optican/demo/model/User.java [hunk 13]
```diff
@@ -16,9 +16,14 @@
...
     @Column(name = "name")
     private String name;
     @Column(name = "password")
     private String password;
+
+    private String email;
+    private String phone;
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/model/User.java:22]: add annotetion Column for all fields

File: src/main/java/com/pharmacy/optican/demo/security/UserDetailsServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Autowired
+    public UserDetailsServiceImpl(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+
+        Optional<User> user = userService.findUserByName(username);
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/security/UserDetailsServiceImpl.java]: use email instead username

File: src/main/java/com/pharmacy/optican/demo/security/UserSecurity.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    private final User user;
+
+    public UserSecurity(User user) {
+        this.user = user;
+    }
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of(()->"ROLE_user");
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/security/UserSecurity.java:20]: use Enum instead hardcoded string with roles

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        this.userRepository = userRepository;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    public void saveUser(User user){
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        userRepository.save(user);
+    }
+
+    public Optional<User> findUserByName(String username){
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/service/UserService.java]: return User instead optional

File: src/main/resources/db/migration/V1_0_0_2__add_user_attributes.sql [hunk 17]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE users
+ADD Email varchar(255),
+ADD Phone varchar(255),
+ADD Firstname varchar(255),
```
Comment (EvgenKalashnikov) [inline src/main/resources/db/migration/V1_0_0_2__add_user_attributes.sql:4]: use cebab case for naming in db

Comment (EvgenKalashnikov) [reply src/main/resources/db/migration/V1_0_0_2__add_user_attributes.sql:4]: for example first_name

File: src/main/resources/templates/registration-page.html [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    <link rel="stylesheet" th:href="@{/css/style.css}">
+</head>
+<body>
+<header class="top">
+    
+</header>
+<nav>
+    <ul>
+        <li><a href="/main-page">ГЛАВНАЯ СТРАНИЦА</a></li>
+        <li sec:authorize="isAnonymous()"><a href="/login-page">ВХОД</a>
```
Comment (EvgenKalashnikov) [inline src/main/resources/templates/registration-page.html]: обьеденить вход и регистрацию

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    public void saveUser(User user){
+        if(validateUser(user)) {
+            user.setPassword(passwordEncoder.encode(user.getPassword()));
+            try {
+                userRepository.save(user);
+                setAuth(user);
+            }catch (Exception e){
+                System.out.println("Ошибка записи пользователя");
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/service/UserService.java:36]: use english for describe exeptons

Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/service/UserService.java:36]: use Log4j instead stdout

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        Optional<User> user = userRepository.findUserByPhone(phone);
+        return user.orElse(null);
+    }
+
+    @PreAuthorize("isAuthenticated()")
+    public void updateUser(User user){
+        Optional<User> oldUser = userRepository.findById(user.getId());
+        oldUser.ifPresent(u -> {
+
+            if(!user.getEmail().isEmpty() && !user.getEmail().equals(u.getEmail())){
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/service/UserService.java:57]: use beanUtils for copy properties

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+            }
+            if(!user.getPassword().isEmpty() && !user.getPassword().equals(u.getPassword())) {
+                u.setPassword(user.getPassword());
+                saveUser(u);
+            } else {
+                    try {
+                        userRepository.save(u);
+                        setAuth(u);
+                    } catch (Exception e){
+                        System.out.println("Ошибка обновления в базу");
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/service/UserService.java:74]: english

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+                u.setFullName(user.getFullName());
+            }
+            if(!user.getPassword().isEmpty() && !user.getPassword().equals(u.getPassword())) {
+                u.setPassword(user.getPassword());
+                saveUser(u);
+            } else {
+                    try {
+                        userRepository.save(u);
+                        setAuth(u);
+                    } catch (Exception e){
```
Comment (EvgenKalashnikov) [inline src/main/java/com/pharmacy/optican/demo/service/UserService.java:73]: catch more exactly exeption

**Review by EvgenKalashnikov — APPROVED** [review]

File: src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java (hunk 2 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java]: done

File: src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java (hunk 3 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java]: done

File: src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java (hunk 1 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/config/SecurityConfig.java:23]: done

File: src/main/java/com/pharmacy/optican/demo/controller/PublicController.java (hunk 4 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/PublicController.java]: done

File: src/main/java/com/pharmacy/optican/demo/controller/PublicController.java (hunk 5 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/PublicController.java]: done

Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/PublicController.java]: done

File: src/main/java/com/pharmacy/optican/demo/security/UserDetailsServiceImpl.java (hunk 14 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/security/UserDetailsServiceImpl.java]: done

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java (hunk 16 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/service/UserService.java]: done

File: src/main/resources/db/migration/V1_0_0_2__add_user_attributes.sql (hunk 17 above)
Comment (Main444) [reply src/main/resources/db/migration/V1_0_0_2__add_user_attributes.sql:4]: done

File: src/main/resources/templates/registration-page.html (hunk 18 above)
Comment (Main444) [reply src/main/resources/templates/registration-page.html]: done

File: src/main/java/com/pharmacy/optican/demo/controller/PublicController.java (hunk 6 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/PublicController.java]: done

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java (hunk 7 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: done

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java (hunk 8 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/UserController.java:26]: done

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java (hunk 9 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: done

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java (hunk 10 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: done

File: src/main/java/com/pharmacy/optican/demo/controller/UserController.java (hunk 11 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/controller/UserController.java]: done

File: src/main/java/com/pharmacy/optican/demo/model/User.java (hunk 12 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/model/User.java]: done

File: src/main/java/com/pharmacy/optican/demo/model/User.java (hunk 13 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/model/User.java:22]: done

File: src/main/java/com/pharmacy/optican/demo/security/UserSecurity.java (hunk 15 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/security/UserSecurity.java:20]: done

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java (hunk 19 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/service/UserService.java:36]: done

Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/service/UserService.java:36]: done

File: src/main/java/com/pharmacy/optican/demo/service/UserService.java (hunk 21 above)
Comment (Main444) [reply src/main/java/com/pharmacy/optican/demo/service/UserService.java:74]: done

**Pull request merged**